gse-afs-ext-base/src/main/java/com/powsybl/gse/afs/ext/base/ContingencyStoreEditor.java

```diff
@@ -45,29 +46,88 @@
     private final class ContingencyTreeCell extends TreeCell<Object> {
 
         private ContingencyTreeCell() {
-            setOnDragDropped(event -> {
-                Dragboard db = event.getDragboard();
-                boolean success = false;
+            setOnDragDetected(this::onDragDetected);
+            setOnDragOver(this::onDragOver);
+            setOnDragDropped(this::onDragDropped);
+        }
+
+        private TreeItem<Object> checkTreeItem() {
+            TreeItem<Object> contingencyItem = null;
+            if (getTreeItem() != null) {
+                if (getTreeItem().getValue() instanceof Contingency) {
+                    contingencyItem = getTreeItem();
+                } else if (getTreeItem().getValue() instanceof ContingencyElement) {
+                    contingencyItem = getTreeItem().getParent();
+                }
+            }
+            return contingencyItem;
+        }
+
+        private EquipmentInfo createEquipmentInfo(ContingencyElement contingencyElement) {
+            ContingencyElementType type = contingencyElement.getType();
+            if (type.equals(ContingencyElementType.BUSBAR_SECTION)) {
+                return new EquipmentInfo(new IdAndName(contingencyElement.getId(), contingencyElement.getId()), "BUSBAR_SECTION");
+            } else if (type.equals(ContingencyElementType.GENERATOR)) {
+                return new EquipmentInfo(new IdAndName(contingencyElement.getId(), contingencyElement.getId()), "GENERATOR");
+            } else if (type.equals(ContingencyElementType.HVDC_LINE)) {
+                return new EquipmentInfo(new IdAndName(contingencyElement.getId(), contingencyElement.getId()), "HVDC_LINE");
+            } else if (type.equals(ContingencyElementType.BRANCH)) {
+                return new EquipmentInfo(new IdAndName(contingencyElement.getId(), contingencyElement.getId()), "BRANCH");
+            } else {
+                return null;
+            }
+        }
+
+        private void onDragDetected(MouseEvent event) {
+            TreeItem<Object> item = contingencyTree.getSelectionModel().getSelectedItem();
+            if (item.getValue() instanceof ContingencyElement) {
+                Dragboard db = startDragAndDrop(TransferMode.COPY_OR_MOVE);
+                ClipboardContent content = new ClipboardContent();
+                content.put(EquipmentInfo.DATA_FORMAT, createEquipmentInfo((ContingencyElement) item.getValue()));
+                db.setContent(content);
+                event.consume();
+            }
+        }
+
+        private void onDragOver(DragEvent event) {
+            Dragboard db = event.getDragboard();
+            if (!(event.getGestureSource() instanceof ContingencyTreeCell) ||
+                    checkTreeItem() != ((ContingencyTreeCell) event.getGestureSource()).checkTreeItem()) {
```

Comment:
Firstly I think this code is quiet difficult to understand. In that case, feel free to put it in a dedicated method with an explicit name to hide complexity.

I think there is only one thing to check here, that can maybe be done in `acceptSingleDraggedElement` & `acceptMultipleDraggedElements`: test if the target node doesn't already contain one of the contingency element you're trying to drop.